pkg/proctree/process.go

```diff
@@ -0,0 +1,375 @@
+package proctree
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/aquasecurity/tracee/pkg/utils/types"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+// processNode is a node in the process tree representing a process.
+// Its purpose is to keep the information of the process, and the connections to other processes
+// and threads nodes.
+// The current implementation is using mutex, so all internal members are assumed to be protected
+// by it (so are not necessarily thread-safe types)
+type processNode struct {
+	// TODO: Add information about the processes like opened files,
+	//  network activities (like TCP connections), argv, environment variables,
+	//  loader and interpreter
+	id             int
+	creationTime   int
+	nsId           int
+	userId         int
+	containerId    string
+	genInfoLock    sync.Once
+	execInfo       *types.TimeSeries[procExecInfo]
+	forkTime       types.Timestamp
+	exitTime       types.Timestamp
+	parentProcess  *processNode
+	childProcesses map[int]*processNode
+	threads        map[int]*threadNode
+	mutex          sync.RWMutex // Protection on all accesses to the process, except for PID reading
+}
+
+// procExecInfo is the information about a process which is changed upon execution
+type procExecInfo struct {
+	Cmd             []string
+	ExecutionBinary FileInfo
+}
+
+func newProcessNode(id int) (*processNode, error) {
+	return &processNode{
+		id:             id,
+		creationTime:   int(time.Now().UnixNano()),
+		threads:        make(map[int]*threadNode),
+		childProcesses: make(map[int]*processNode),
+		execInfo:       types.NewTimeSeries[procExecInfo](procExecInfo{}),
+	}, nil
+}
+
+// setGenInfoFromEventProtected fill the general info of the process (information
+// of the event given by every process received from it) in a way that it is filled
+// only once in an efficient way to reduce performance penalty.
+// This method uses the process lock to reduce unnecessary locking if not needed. Make sure to unlock
+// the lock before using this.
+func (p *processNode) setGenInfoFromEventProtected(event *trace.Event) {
+	p.genInfoLock.Do(
+		func() {
+			p.mutex.Lock()
+			p.fillGeneralInfo(
+				event.ProcessID,
+				event.UserID,
+				event.ContainerID,
+			)
+			p.mutex.Unlock()
+		},
+	)
+}
+
+// setGeneralInfoProtected is used to fill general information of a process
+// only once, but when the information cannot be retrieved from an event like
+// with setGenInfoFromEventProtected.
+// This method is not protected by locks.
+func (p *processNode) setGeneralInfoProtected(
```

Comment:
It does not make any sense to me to have protected and unprotected versions of the same change to internal fields. Also, if you already have a "Once" guaranteeing that the operation will be done just once, why not to have only a protected version ? There is no free lunch when it comes to concurrency (its either an atomic change or a protected (mutex,futex,spins) transactional change). We can't mix both.

Reply:
So the reason here is that I need to lock the node when applying the changes.
I wanted to lock the node from inside the `Once` function, to avoid locking the node for nothing after the first call.
However, when processing fork event, I lock the mutex already.
I don't want to lock it twice, so I created another function to fill the data if the mutex is locked from outside.
What do you propose to do here to avoid unnecessary locking twice in fork?